### PR TITLE
feat: ensure proportional distribution of prs

### PR DESCRIPTION
**Pull request by joe-bowman**
## 1. Summary
Fixes QCK-152

Distributes participation rewards proportional to total qasset supply.

## 2.Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## 3. Implementation details

Previously participation rewards were distributed based upon the user's claims as a proportion of all claims made for a given asset. This results is disproportionate amounts of assets being distributes in the event that very few claims are made in a given epoch.

Instead we distribute as a proportion of a user's claims total as a percentage of the qassets in supply, thus ensuring that users are rewards appropriately.

The remaining tokens are returned to the incentives pool.

## 4. How to test/use

Make claims for holding tokens; rewards received should be proportional to the users' claim vs. total supply.

### CODE REVIEW

File: x/participationrewards/keeper/keeper.go [hunk 1]
```diff
@@ -21,10 +21,10 @@ import (
...
 
-// userAllocation is an internal keeper struct to track transient state for
+// UserAllocation is an internal keeper struct to track transient state for
 // rewards distribution. It contains the user address and the coins that are
 // allocated to it.
-type userAllocation struct {
+type UserAllocation struct {
 	Address string
 	Amount  math.Int
 }
```
Comment (aljo242) [inline x/participationrewards/keeper/keeper.go:30]: We should move this to` types` now that it is exported.

File: x/participationrewards/keeper/keeper.go [hunk 2]
```diff
@@ -121,7 +121,7 @@ func (k Keeper) UpdateSelfConnectionData(ctx sdk.Context) error {
 	selfConnectionData, err := json.Marshal(types.ConnectionProtocolData{
 		ConnectionID: types.SelfConnection,
 		ChainID:      ctx.ChainID(),
-		LastEpoch:    ctx.BlockHeight() - 1,
+		LastEpoch:    ctx.BlockHeight() - 2, // reason why -2 works here.
```
Comment (aljo242) [inline x/participationrewards/keeper/keeper.go:124]: What does this comment mean?

File: x/participationrewards/keeper/rewards_holdings.go [hunk 3]
```diff
@@ -5,55 +5,59 @@ import (
...
 	"github.com/ingenuity-build/quicksilver/utils"
+	airdroptypes "github.com/ingenuity-build/quicksilver/x/airdrop/types"
 	cmtypes "github.com/ingenuity-build/quicksilver/x/claimsmanager/types"
 	icstypes "github.com/ingenuity-build/quicksilver/x/interchainstaking/types"
+	"github.com/ingenuity-build/quicksilver/x/participationrewards/types"
 )
 
-func (k Keeper) allocateHoldingsRewards(ctx sdk.Context) error {
+func (k Keeper) AllocateHoldingsRewards(ctx sdk.Context) error {
 	k.Logger(ctx).Info("allocateHoldingsRewards")
```
Comment (aljo242) [inline x/participationrewards/keeper/rewards_holdings.go:15]:
```suggestion
	k.Logger(ctx).Info("AllocateHoldingsRewards")
```

Comment (aljo242) [reply x/participationrewards/keeper/rewards_holdings.go:15]: if we still want to keep

File: x/participationrewards/keeper/rewards_validatorSelection.go [hunk 4]
```diff
@@ -86,21 +87,21 @@ func (k Keeper) getZoneScores(
...
 		return nil, err
 	}
 
 	return &zs, nil
 }
 
-// calcDistributionScores calculates the validator distribution scores for the
+// CalcDistributionScores calculates the validator distribution scores for the
 // given zone based on the normalized voting power of the validators; scoring
 // favours smaller validators for decentraliztion purposes.
```
Comment (aljo242) [inline x/participationrewards/keeper/rewards_validatorSelection.go:103]:
```suggestion
// favours smaller validators for decentralization purposes.
```

**Review by aljo242 — COMMENTED** [review]
Newly exported functions should be tested:
- `AllocateHoldingsRewards()`
- `AllocateLockupRewards()`
- `AllocateValidatorSelectionRewards()`
- `CalcDistributionScores()`
- `CalcOverallScores()`
- `CalcUserValidatorSelectionAllocations()`

We should add or create an issue to track adding them in the future

File: x/participationrewards/keeper/keeper.go (hunk 2 above)
Comment (ajansari95) [reply x/participationrewards/keeper/keeper.go:124]: so this means to investigate the reason of off by 1.

File: x/participationrewards/keeper/keeper.go (hunk 1 above)
Comment (ajansari95) [reply x/participationrewards/keeper/keeper.go:30]: second that

Comment (joe-bowman) [reply x/participationrewards/keeper/keeper.go:30]: done

**Comment by joe-bowman** [comment]
I believe this is ready for review/merge into develop now.

**Review by ajansari95 — APPROVED** [review]
LGTM

**Review by aljo242 — APPROVED** [review]

**Pull request merged**